blockchain/utxoviewpoint.go

```diff
@@ -454,15 +461,68 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
 	return nil
 }
 
-// disconnectTransactions updates the view by removing all utxos created by
-// the transactions in either the regular or stake tree of the block, depending
-// on the flag, and unspending all of the txos spent by those same transactions
-// by using the provided spent txo information.
-func (view *UtxoViewpoint) disconnectTransactions(block *dcrutil.Block, stxos []spentTxOut, stakeTree bool, isTreasuryEnabled bool) error {
+// fetchEntryByHash attempts to find any available utxo for the given hash by
+// searching the entire set of possible outputs for the given hash.  It checks
+// the view first and then falls back to the database if needed.
+//
+// It should be noted that this is quite inefficient, and is currently only
+// called when disconnecting a transaction and the spent transaction output does
+// not have all of the information that is needed.  That is only the case for
+// the legacy v2 spend journal format, which only stored the containing
+// transaction info when the output was the last unspent output of the
+// transaction.  Realistically, that case will almost never run since all new
+// entries include the information for all outputs and thus the only way this
+// will be hit is if a long enough reorg happens such that a block with the old
+// spend data is being disconnected.  The probability of that in practice is
+// extremely low to begin with and becomes vanishingly small the more new blocks
+// are connected.  In the case of a fresh database that has only ever run with
+// the new v2 format, that case would never be run.
+func (view *UtxoViewpoint) fetchEntryByHash(db database.DB, hash *chainhash.Hash) (*UtxoEntry, error) {
+	// First attempt to find a utxo with the provided hash in the view. Note that
+	// this is only a best effort and will fallback to check the database as well
+	// if the utxo is not found in the view.
+	//
+	// The first three outputs are checked because it covers the majority of
+	// common scenarios, including:
+	//   - The vast majority of regular transactions consist of two outputs where
+	//     one is some form of "pay-to-somebody-else" and the other is a change
+	//     output.
+	//   - For some stake transactions (e.g. votes) the first two outputs will
+	//     never be found as a utxo, since they are not spendable, so we need to
+	//     check through at least the third output.
+	trees := []int8{wire.TxTreeRegular, wire.TxTreeStake}
+	for _, tree := range trees {
+		outpoint := wire.OutPoint{Hash: *hash, Tree: tree}
+		for idx := uint32(0); idx < 3; idx++ {
```

Comment:
Good catch in regards to the entry loaded from the database being unmodified.  I Will update this to check all possible outputs here.

Reply:
This has been updated to use a pre-calculated `maxLegacyOutputsPerBlock` const to check all possible outputs.

As already mentioned in the function and inline comments, this is super inefficient but should really never happen.  Also, in the case it should happen, it will return immediately after finding the output in the view (which will be an index much lower than `maxLegacyOutputsPerBlock` and would only go to `maxLegacyOutputsPerBlock` in the worst possible case or if the output does not exist in the view).

As already mentioned in the PR description, we would need to migrate the data to avoid this, but there are some complications with attempting to do that efficiently.  One alternative to migrating would be to iterate all blocks and create the spend journal entries from scratch in the new format.  However, this requires context checks, such as whether the treasury agenda is active to determine the tx type, so it would be slightly different from current migrations and would have to run after the block index is loaded. I think this will probably end up being the best approach, and I can follow up on it separately from this PR.